internal/vulnstore/postgres/enrichment.go

```diff
@@ -0,0 +1,224 @@
+INSERT
+INTO
+	uo_enrich (enrich, updater, uo, date)
+VALUES
+	(
+		(
+			SELECT
+				id
+			FROM
+				enrichment
+			WHERE
+				hash_kind = $1
+				AND hash = $2
+				AND updater = $3
+		),
+		$3,
+		$4,
+		transaction_timestamp()
+	)
+ON CONFLICT
+DO
+	NOTHING;`
+	)
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "internal/vulnstore/postgres/UpdateEnrichments"))
+
+	tx, err := s.pool.Begin(ctx)
+	if err != nil {
+		return uuid.Nil, fmt.Errorf("unable to start transaction: %w", err)
+	}
+	defer tx.Rollback(ctx)
+
+	var id uint64
+	var ref uuid.UUID
+
+	start := time.Now()
+
+	if err := s.pool.QueryRow(ctx, create, name, string(fp)).Scan(&id, &ref); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to create update_operation: %w", err)
+	}
+
+	updateEnrichmentsCounter.WithLabelValues("create").Add(1)
+	updateEnrichmentsDuration.WithLabelValues("create").Observe(time.Since(start).Seconds())
+
+	zlog.Debug(ctx).
+		Str("ref", ref.String()).
+		Msg("update_operation created")
+
+	batch := microbatch.NewInsert(tx, 2000, time.Minute)
+	start = time.Now()
+	for i := range es {
+		hashKind, hash := hashEnrichment(&es[i])
+		err := batch.Queue(ctx, insert,
+			hashKind, hash, name, es[i].Tags, es[i].Enrichment,
+		)
+		if err != nil {
+			return uuid.Nil, fmt.Errorf("failed to queue enrichment: %w", err)
+		}
+		if err := batch.Queue(ctx, assoc, hashKind, hash, name, id); err != nil {
+			return uuid.Nil, fmt.Errorf("failed to queue association: %w", err)
+		}
+	}
+	if err := batch.Done(ctx); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to finish batch enrichment insert: %w", err)
+	}
+	updateEnrichmentsCounter.WithLabelValues("insert_batch").Add(1)
+	updateEnrichmentsDuration.WithLabelValues("insert_batch").Observe(time.Since(start).Seconds())
+
+	if err := tx.Commit(ctx); err != nil {
+		return uuid.Nil, fmt.Errorf("failed to commit transaction: %w", err)
+	}
+	zlog.Debug(ctx).
+		Stringer("ref", ref).
+		Int("inserted", len(es)).
+		Msg("update_operation committed")
+	return ref, nil
+}
+
+func hashEnrichment(r *driver.EnrichmentRecord) (k string, d []byte) {
+	h := md5.New()
+	sort.Strings(r.Tags)
+	for _, t := range r.Tags {
+		io.WriteString(h, t)
+		h.Write([]byte("\x00"))
+	}
+	h.Write(r.Enrichment)
+	return "md5", h.Sum(nil)
+}
+
+func (s *Store) GetEnrichment(ctx context.Context, name string, tags []string) ([]driver.EnrichmentRecord, error) {
+	const query = `
+WITH
+	latest
+		AS (
+			SELECT
+				max(id) AS id
+			FROM
+				update_operation
+			WHERE
+				updater = $1
+		)
+SELECT
+	e.tags, e.data
+FROM
+	enrichment AS e,
+	uo_enrich AS uo,
+	latest
+WHERE
+	uo.uo = latest.id
+	AND uo.enrich = e.id
+	AND e.tags && $2::text[];`
+
+	ctx = baggage.ContextWithValues(ctx,
+		label.String("component", "internal/vulnstore/postgres/GetEnrichment"))
+	tx, err := s.pool.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback(ctx)
+
+	results := make([]driver.EnrichmentRecord, 0, 8) // Guess at capacity.
```

Comment:
Is 8 based on something you've seen in the wild?

Reply:
Nope